File: bindings/lua/Polycode/Mesh.lua
require "Polycode/Resource"

class "Mesh" (Resource)


function Mesh:Mesh(...)
	local arg = {...}
	if type(arg[1]) == "table" and count(arg) == 1 then
		if ""..arg[1].__classname == "Resource" then
			self.__ptr = arg[1].__ptr
			return
		end
	end
	for k,v in pairs(arg) do
		if type(v) == "table" then
			if v.__ptr ~= nil then
				arg[k] = v.__ptr
			end
		end
	end
	if self.__ptr == nil and arg[1] ~= "__skip_ptr__" then
		self.__ptr = Polycode.Mesh(unpack(arg))
	end
end

function Mesh:loadMesh(fileName)
	local retVal = Polycode.Mesh_loadMesh(self.__ptr, fileName)
end

function Mesh:saveToFile(fileName, writeNormals, writeTangents, writeColors, writeBoneWeights, writeUVs, writeSecondaryUVs)
	local retVal = Polycode.Mesh_saveToFile(self.__ptr, fileName, writeNormals, writeTangents, writeColors, writeBoneWeights, writeUVs, writeSecondaryUVs)
end

function Mesh:addSubmesh(newSubmesh)
	local retVal = Polycode.Mesh_addSubmesh(self.__ptr, newSubmesh.__ptr)
end

function Mesh:removeSubmeshAtIndex(index)
	local retVal = Polycode.Mesh_removeSubmeshAtIndex(self.__ptr, index)
end

function Mesh:getNumSubmeshes()
	local retVal =  Polycode.Mesh_getNumSubmeshes(self.__ptr)
	return retVal
end

function Mesh:getSubmeshAtIndex(index)
	local retVal = Polycode.Mesh_getSubmeshAtIndex(self.__ptr, index)
	if retVal == nil then return nil end
	local __c = _G["MeshGeometry"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function Mesh:getSubmeshPointer(index)
	local retVal = Polycode.Mesh_getSubmeshPointer(self.__ptr, index)
	if retVal == nil then return nil end
	local __c = _G["shared_ptr<MeshGeometry>"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function Mesh:clearMesh()
	local retVal =  Polycode.Mesh_clearMesh(self.__ptr)
end

function Mesh:calculateBBox()
	local retVal =  Polycode.Mesh_calculateBBox(self.__ptr)
	if retVal == nil then return nil end
	local __c = _G["Vector3"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function Mesh:getRadius()
	local retVal =  Polycode.Mesh_getRadius(self.__ptr)
	return retVal
end

function Mesh:__delete()
	if self then Polycode.delete_Mesh(self.__ptr) end
end

File: bindings/javascript/Polycode/Mesh.js
require('Polycode/Resource')

function Mesh() {
	if(arguments[0] != "__skip_ptr__") {
		this.__ptr = Polycode.Mesh()
	}
}

Mesh.prototype = Object.create(Resource.prototype);


Mesh.prototype.MeshFromFileName = function(fileName) {
	var retVal = new Mesh()
	retVal.__ptr = Polycode.Mesh_MeshFromFileName(fileName)
	return retVal
}

Mesh.prototype.Copy = function() {
	var retVal = new Mesh()
	retVal.__ptr = Polycode.Mesh_Copy(this.__ptr)
	return retVal
}

Mesh.prototype.loadMesh = function(fileName) {
	Polycode.Mesh_loadMesh(this.__ptr, fileName)
}

Mesh.prototype.saveToFile = function(fileName,writeNormals,writeTangents,writeColors,writeBoneWeights,writeUVs,writeSecondaryUVs) {
	Polycode.Mesh_saveToFile(this.__ptr, fileName, writeNormals, writeTangents, writeColors, writeBoneWeights, writeUVs, writeSecondaryUVs)
}

Mesh.prototype.loadFromFile = function(inFile) {
	Polycode.Mesh_loadFromFile(this.__ptr, inFile.__ptr)
}

Mesh.prototype.addSubmesh = function(newSubmesh) {
	Polycode.Mesh_addSubmesh(this.__ptr, newSubmesh)
}

Mesh.prototype.removeSubmeshAtIndex = function(index) {
	Polycode.Mesh_removeSubmeshAtIndex(this.__ptr, index)
}

Mesh.prototype.getNumSubmeshes = function() {
	return Polycode.Mesh_getNumSubmeshes(this.__ptr)
}

Mesh.prototype.getSubmeshAtIndex = function(index) {
	var retVal = new MeshGeometry()
	retVal.__ptr = Polycode.Mesh_getSubmeshAtIndex(this.__ptr, index)
	return retVal
}

Mesh.prototype.getSubmeshPointer = function(index) {
	var retVal = new MeshGeometry()
	retVal.__ptr = Polycode.Mesh_getSubmeshPointer(this.__ptr, index)
	return retVal
}

Mesh.prototype.clearMesh = function() {
	Polycode.Mesh_clearMesh(this.__ptr)
}

Mesh.prototype.calculateBBox = function() {
	var retVal = new Vector3()
	retVal.__ptr = Polycode.Mesh_calculateBBox(this.__ptr)
	return retVal
}

Mesh.prototype.getRadius = function() {
	return Polycode.Mesh_getRadius(this.__ptr)
}
